### PR TITLE
Colour graph by napari track colur

### PR DESCRIPTION
**Pull request by dstansby**
Fixes https://github.com/lowe-lab-ucl/arboretum/issues/9.

### CODE REVIEW

File: napari_arboretum/graph.py [hunk 1]
```diff
@@ -117,7 +117,7 @@ def build_subgraph(layer, node):
     -------
     root_id : int, None
         The root node ID of the tree which contains the node.
-    nodes : list
+    nodes : list[TreeNode]
```
Comment (paddyroddy) [inline napari_arboretum/graph.py:120]: can you add the appropriate typing above? I think here it would be `tuple[Union[Optional[int]], list[TreeNode]]`

Comment (paddyroddy) [reply napari_arboretum/graph.py:120]: you've added it in docstring but not the signature yeah (in my view it only needs to be one, but I think the signature nicer for `mypy`)

Comment (dstansby) [reply napari_arboretum/graph.py:122]: 👍 done

**Comment by paddyroddy** [comment]
sorry based on this morning's discussion I need to work out how to run this before I carry on with my review

File: napari_arboretum/plugin.py [hunk 2]
```diff
@@ -84,41 +87,71 @@ def _append_mouse_callback(self, layer: napari.layers.Tracks):
...
+        Update colors on a list of edges from the colors they have in a
+        track layer. Note that this updates the colors in-place on ``edges``.
+        """
+        for e in self.edges:
+            if e.id is not None:
+                color = self.layer.track_colors[
+                    np.where(self.layer.properties["track_id"] == e.id)
+                ][-1]
+                # napari uses [0, 1] RGBA, pygraphqt uses [0, 255] RGBA
+                e.color = color * 256 - 1
```
[the remark this answers is not in the document]
Comment (dstansby) [reply napari_arboretum/plugin.py:119]: Of course 🤦, thanks for catching

File: napari_arboretum/plugin.py [hunk 3]
```diff
@@ -84,41 +87,71 @@ def _append_mouse_callback(self, layer: napari.layers.Tracks):
...
+    def update_colors(self):
+        """
+        Update colors on a list of edges from the colors they have in a
+        track layer. Note that this updates the colors in-place on ``edges``.
+        """
+        for e in self.edges:
+            if e.id is not None:
+                color = self.layer.track_colors[
+                    np.where(self.layer.properties["track_id"] == e.id)
+                ][-1]
```
Comment (paddyroddy) [inline napari_arboretum/plugin.py:117]: would `self.layer.track_colors[self.layer.properties["track_id"] == e.id][-1]` achieve the same thing?

File: napari_arboretum/tree.py [hunk 4]
```diff
@@ -7,6 +10,14 @@
 RED = np.array([255, 0, 0, 255], dtype=np.uint8)
 
 
+@dataclass
+class Edge:
+    x: Tuple[float, float]
+    y: Tuple[float, float]
+    color: Tuple[float, float, float, float]
+    id: Union[int, None] = None
```
Comment (paddyroddy) [inline napari_arboretum/tree.py:18]: This is equivalent but the preferred way, I believe. Will need to import `Optional` from `typing`

```suggestion
    id: Optional[int] = None
```

Comment (paddyroddy) [inline napari_arboretum/tree.py:18]: This is equivalent but the preferred way, I believe. Will need to import `Optional` from `typing`

```suggestion
    id: Optional[int] = None
```

**Review by paddyroddy — APPROVED** [review]
👍

**Comment by dstansby** [comment]
Since this has an approval and the linux tests pass, I'm going to merge. https://github.com/lowe-lab-ucl/arboretum/pull/35 needs to be merged to fix the windows tests.

**Pull request merged**